File: dkplayer-players/exo/src/main/java/com/dueeeke/videoplayer/exo/ExoMediaPlayer.java
package com.dueeeke.videoplayer.exo;

import android.content.Context;
import android.content.res.AssetFileDescriptor;
import android.os.Handler;
import android.view.Surface;
import android.view.SurfaceHolder;

import com.dueeeke.videoplayer.player.AbstractPlayer;
import com.dueeeke.videoplayer.player.VideoViewManager;
import com.google.android.exoplayer2.DefaultLoadControl;
import com.google.android.exoplayer2.DefaultRenderersFactory;
import com.google.android.exoplayer2.ExoPlaybackException;
import com.google.android.exoplayer2.ExoPlayerFactory;
import com.google.android.exoplayer2.LoadControl;
import com.google.android.exoplayer2.PlaybackParameters;
import com.google.android.exoplayer2.Player;
import com.google.android.exoplayer2.RenderersFactory;
import com.google.android.exoplayer2.SimpleExoPlayer;
import com.google.android.exoplayer2.source.DefaultMediaSourceEventListener;
import com.google.android.exoplayer2.source.MediaSource;
import com.google.android.exoplayer2.source.MediaSourceEventListener;
import com.google.android.exoplayer2.trackselection.DefaultTrackSelector;
import com.google.android.exoplayer2.trackselection.TrackSelector;
import com.google.android.exoplayer2.util.Log;
import com.google.android.exoplayer2.video.VideoListener;

import java.util.Map;


public class ExoMediaPlayer extends AbstractPlayer implements VideoListener, Player.EventListener {

    protected Context mAppContext;
    protected SimpleExoPlayer mInternalPlayer;
    protected MediaSource mMediaSource;
    protected ExoMediaSourceHelper mMediaSourceHelper;

    private PlaybackParameters mSpeedPlaybackParameters;

    private int mLastReportedPlaybackState = Player.STATE_IDLE;
    private boolean mLastReportedPlayWhenReady = false;
    private boolean mIsPreparing;
    private boolean mIsBuffering;

    private LoadControl mLoadControl;
    private RenderersFactory mRenderersFactory;
    private TrackSelector mTrackSelector;

    public ExoMediaPlayer(Context context) {
        mAppContext = context.getApplicationContext();
        mMediaSourceHelper = ExoMediaSourceHelper.getInstance(context);
    }

    @Override
    public void initPlayer() {
        mLoadControl = mLoadControl == null ? new DefaultLoadControl() : mLoadControl;
        mRenderersFactory = mRenderersFactory == null ? new DefaultRenderersFactory(mAppContext) : mRenderersFactory;
        mTrackSelector = mTrackSelector == null ? new DefaultTrackSelector() : mTrackSelector;
        mInternalPlayer = ExoPlayerFactory.newSimpleInstance(mAppContext, mRenderersFactory, mTrackSelector, mLoadControl);
        setOptions();
        Log.setLogLevel(VideoViewManager.getConfig().mIsEnableLog ? Log.LOG_LEVEL_ALL : Log.LOG_LEVEL_OFF);
        Log.setLogStackTraces(VideoViewManager.getConfig().mIsEnableLog);
        mInternalPlayer.addListener(this);
        mInternalPlayer.addVideoListener(this);
    }

    public void setTrackSelector(TrackSelector trackSelector) {
        mTrackSelector = trackSelector;
    }

    public void setRenderersFactory(RenderersFactory renderersFactory) {
        mRenderersFactory = renderersFactory;
    }

    public void setLoadControl(LoadControl loadControl) {
        mLoadControl = loadControl;
    }

    @Override
    public void setDataSource(String path, Map<String, String> headers) {
        mMediaSource = mMediaSourceHelper.getMediaSource(path, headers);
    }

    @Override
    public void setDataSource(AssetFileDescriptor fd) {
        //no support
    }

    @Override
    public void start() {
        if (mInternalPlayer == null)
            return;
        mInternalPlayer.setPlayWhenReady(true);
    }

    @Override
    public void pause() {
        if (mInternalPlayer == null)
            return;
        mInternalPlayer.setPlayWhenReady(false);
    }

    @Override
    public void stop() {
        if (mInternalPlayer == null)
            return;
        mInternalPlayer.stop();
    }

    @Override
    public void prepareAsync() {
        if (mInternalPlayer == null)
            return;
        if (mMediaSource == null) return;
        if (mSpeedPlaybackParameters != null) {
            mInternalPlayer.setPlaybackParameters(mSpeedPlaybackParameters);
        }
        mIsPreparing = true;
        mMediaSource.addEventListener(new Handler(), mMediaSourceEventListener);
        mInternalPlayer.prepare(mMediaSource);
    }

    private MediaSourceEventListener mMediaSourceEventListener = new DefaultMediaSourceEventListener() {
        @Override
        public void onReadingStarted(int windowIndex, MediaSource.MediaPeriodId mediaPeriodId) {
            super.onReadingStarted(windowIndex, mediaPeriodId);
            if (mPlayerEventListener != null && mIsPreparing) {
                mPlayerEventListener.onPrepared();
            }
        }
    };

    @Override
    public void reset() {
        if (mInternalPlayer != null) {
            mInternalPlayer.stop(true);
            mInternalPlayer.setVideoSurface(null);
            mIsPreparing = false;
            mIsBuffering = false;
            mLastReportedPlaybackState = Player.STATE_IDLE;
            mLastReportedPlayWhenReady = false;
        }
    }

    @Override
    public boolean isPlaying() {
        if (mInternalPlayer == null)
            return false;
        int state = mInternalPlayer.getPlaybackState();
        switch (state) {
            case Player.STATE_BUFFERING:
            case Player.STATE_READY:
                return mInternalPlayer.getPlayWhenReady();
            case Player.STATE_IDLE:
            case Player.STATE_ENDED:
            default:
                return false;
        }
    }

    @Override
    public void seekTo(long time) {
        if (mInternalPlayer == null)
            return;
        mInternalPlayer.seekTo(time);
    }

    @Override
    public void release() {
        if (mInternalPlayer != null) {
            mInternalPlayer.removeListener(this);
            mInternalPlayer.removeVideoListener(this);
            final SimpleExoPlayer player = mInternalPlayer;
            mInternalPlayer = null;
            new Thread() {
                @Override
                public void run() {
                    //异步释放，防止卡顿
                    player.release();
                }
            }.start();
        }

        mIsPreparing = false;
        mIsBuffering = false;
        mLastReportedPlaybackState = Player.STATE_IDLE;
        mLastReportedPlayWhenReady = false;
        mSpeedPlaybackParameters = null;
    }

    @Override
    public long getCurrentPosition() {
        if (mInternalPlayer == null)
            return 0;
        return mInternalPlayer.getCurrentPosition();
    }

    @Override
    public long getDuration() {
        if (mInternalPlayer == null)
            return 0;
        return mInternalPlayer.getDuration();
    }

    @Override
    public int getBufferedPercentage() {
        return mInternalPlayer == null ? 0 : mInternalPlayer.getBufferedPercentage();
    }

    @Override
    public void setSurface(Surface surface) {
        if (mInternalPlayer != null) {
            mInternalPlayer.setVideoSurface(surface);
        }
    }

    @Override
    public void setDisplay(SurfaceHolder holder) {
        if (holder == null)
            setSurface(null);
        else
            setSurface(holder.getSurface());
    }

    @Override
    public void setVolume(float leftVolume, float rightVolume) {
        if (mInternalPlayer != null)
            mInternalPlayer.setVolume((leftVolume + rightVolume) / 2);
    }

    @Override
    public void setLooping(boolean isLooping) {
        if (mInternalPlayer != null)
            mInternalPlayer.setRepeatMode(isLooping ? Player.REPEAT_MODE_ALL : Player.REPEAT_MODE_OFF);
    }

    @Override
    public void setOptions() {
        //准备好就开始播放
        mInternalPlayer.setPlayWhenReady(true);
    }

    @Override
    public void setSpeed(float speed) {
        PlaybackParameters playbackParameters = new PlaybackParameters(speed);
        mSpeedPlaybackParameters = playbackParameters;
        if (mInternalPlayer != null) {
            mInternalPlayer.setPlaybackParameters(playbackParameters);
        }
    }

    @Override
    public long getTcpSpeed() {
        // no support
        return 0;
    }

    @Override
    public void onPlayerStateChanged(boolean playWhenReady, int playbackState) {
        if (mPlayerEventListener == null) return;
        if (mIsPreparing) return;
        if (mLastReportedPlayWhenReady != playWhenReady || mLastReportedPlaybackState != playbackState) {
            switch (playbackState) {
                case Player.STATE_BUFFERING:
                    mPlayerEventListener.onInfo(MEDIA_INFO_BUFFERING_START, getBufferedPercentage());
                    mIsBuffering = true;
                    break;
                case Player.STATE_READY:
                    if (mIsBuffering) {
                        mPlayerEventListener.onInfo(MEDIA_INFO_BUFFERING_END, getBufferedPercentage());
                        mIsBuffering = false;
                    }
                    break;
                case Player.STATE_ENDED:
                    mPlayerEventListener.onCompletion();
                    break;
            }
            mLastReportedPlaybackState = playbackState;
            mLastReportedPlayWhenReady = playWhenReady;
        }
    }

    @Override
    public void onPlayerError(ExoPlaybackException error) {
        if (mPlayerEventListener != null) {
            mPlayerEventListener.onError();
        }
    }

    @Override
    public void onVideoSizeChanged(int width, int height, int unappliedRotationDegrees, float pixelWidthHeightRatio) {
        if (mPlayerEventListener != null) {
            mPlayerEventListener.onVideoSizeChanged(width, height);
            if (unappliedRotationDegrees > 0) {
                mPlayerEventListener.onInfo(MEDIA_INFO_VIDEO_ROTATION_CHANGED, unappliedRotationDegrees);
            }
        }
    }

    @Override
    public void onRenderedFirstFrame() {
        if (mPlayerEventListener != null && mIsPreparing) {
            mPlayerEventListener.onInfo(MEDIA_INFO_VIDEO_RENDERING_START, 0);
            mIsPreparing = false;
        }
    }
}

File: dkplayer-players/exo/src/main/java/xyz/doikki/videoplayer/exo/ExoMediaPlayer.kt
package xyz.doikki.videoplayer.exo

import android.content.Context
import android.content.res.AssetFileDescriptor
import android.view.Surface
import android.view.SurfaceHolder
import com.google.android.exoplayer2.*
import com.google.android.exoplayer2.analytics.DefaultAnalyticsCollector
import com.google.android.exoplayer2.source.DefaultMediaSourceFactory
import com.google.android.exoplayer2.source.MediaSource
import com.google.android.exoplayer2.trackselection.DefaultTrackSelector
import com.google.android.exoplayer2.trackselection.MappingTrackSelector
import com.google.android.exoplayer2.trackselection.TrackSelector
import com.google.android.exoplayer2.upstream.DefaultBandwidthMeter
import com.google.android.exoplayer2.util.Clock
import com.google.android.exoplayer2.util.EventLogger
import com.google.android.exoplayer2.video.VideoSize
import xyz.doikki.videoplayer.player.AbstractPlayer
import xyz.doikki.videoplayer.player.IPlayer
import xyz.doikki.videoplayer.GlobalConfig
import xyz.doikki.videoplayer.internal.PlayerException

open class ExoMediaPlayer(context: Context) : AbstractPlayer(), Player.Listener {
    protected var mAppContext: Context
    protected var mInternalPlayer: ExoPlayer? = null
    @JvmField
    protected var mMediaSource: MediaSource? = null
    protected var mMediaSourceHelper: ExoMediaSourceHelper
    private var mSpeedPlaybackParameters: PlaybackParameters? = null
    private var mIsPreparing = false
    private var mLoadControl: LoadControl? = null
    private var mRenderersFactory: RenderersFactory? = null
    private var mTrackSelector: TrackSelector? = null
    override fun init() {
        mInternalPlayer = ExoPlayer.Builder(
            mAppContext,
            (if (mRenderersFactory == null) DefaultRenderersFactory(mAppContext).also {
                mRenderersFactory = it
            } else mRenderersFactory)!!,
            DefaultMediaSourceFactory(mAppContext),
            (if (mTrackSelector == null) DefaultTrackSelector(mAppContext).also {
                mTrackSelector = it
            } else mTrackSelector)!!,
            (if (mLoadControl == null) DefaultLoadControl().also {
                mLoadControl = it
            } else mLoadControl)!!,
            DefaultBandwidthMeter.getSingletonInstance(mAppContext),
            DefaultAnalyticsCollector(Clock.DEFAULT))
            .build()
        //准备好就开始播放
        mInternalPlayer!!.playWhenReady = true

        //播放器日志
        if (GlobalConfig.isDebuggable && mTrackSelector is MappingTrackSelector) {
            mInternalPlayer!!.addAnalyticsListener(
                EventLogger(
                    mTrackSelector as MappingTrackSelector?,
                    "ExoPlayer"
                )
            )
        }
        mInternalPlayer!!.addListener(this)
    }

    fun setTrackSelector(trackSelector: TrackSelector?) {
        mTrackSelector = trackSelector
    }

    fun setRenderersFactory(renderersFactory: RenderersFactory?) {
        mRenderersFactory = renderersFactory
    }

    fun setLoadControl(loadControl: LoadControl?) {
        mLoadControl = loadControl
    }

    override fun setDataSource(path: String, headers: Map<String, String>?) {
        mMediaSource = mMediaSourceHelper.getMediaSource(path, headers)
    }

    override fun setDataSource(fd: AssetFileDescriptor) {
        //no support
    }

    override fun start() {
        if (mInternalPlayer == null) return
        mInternalPlayer!!.playWhenReady = true
    }

    override fun pause() {
        if (mInternalPlayer == null) return
        mInternalPlayer!!.playWhenReady = false
    }

    override fun stop() {
        if (mInternalPlayer == null) return
        mInternalPlayer!!.stop()
    }

    override fun prepareAsync() {
        if (mInternalPlayer == null) return
        if (mMediaSource == null) return
        if (mSpeedPlaybackParameters != null) {
            mInternalPlayer!!.playbackParameters = mSpeedPlaybackParameters!!
        }
        mIsPreparing = true
        mInternalPlayer!!.setMediaSource(mMediaSource!!)
        mInternalPlayer!!.prepare()
    }

    override fun reset() {
        if (mInternalPlayer != null) {
            mInternalPlayer!!.stop()
            mInternalPlayer!!.clearMediaItems()
            mInternalPlayer!!.setVideoSurface(null)
            mIsPreparing = false
        }
    }

    override fun isPlaying(): Boolean {
        if (mInternalPlayer == null) return false
        val state = mInternalPlayer!!.playbackState
        return when (state) {
            Player.STATE_BUFFERING, Player.STATE_READY -> mInternalPlayer!!.playWhenReady
            Player.STATE_IDLE, Player.STATE_ENDED -> false
            else -> false
        }
    }

    override fun seekTo(msec: Long) {
        if (mInternalPlayer == null) return
        mInternalPlayer!!.seekTo(msec)
    }

    override fun release() {
        if (mInternalPlayer != null) {
            mInternalPlayer!!.removeListener(this)
            mInternalPlayer!!.release()
            mInternalPlayer = null
        }
        mIsPreparing = false
        mSpeedPlaybackParameters = null
    }

    override fun getCurrentPosition(): Long {
        return if (mInternalPlayer == null) 0 else mInternalPlayer!!.currentPosition
    }

    override fun getDuration(): Long {
        return if (mInternalPlayer == null) 0 else mInternalPlayer!!.duration
    }

    override fun getBufferedPercentage(): Int {
        return if (mInternalPlayer == null) 0 else mInternalPlayer!!.bufferedPercentage
    }

    override fun setSurface(surface: Surface?) {
        if (mInternalPlayer != null) {
            mInternalPlayer!!.setVideoSurface(surface)
        }
    }

    override fun setDisplay(holder: SurfaceHolder?) {
        if (holder == null) setSurface(null) else setSurface(holder.surface)
    }

    override fun setVolume(leftVolume: Float, rightVolume: Float) {
        if (mInternalPlayer != null) mInternalPlayer!!.volume = (leftVolume + rightVolume) / 2
    }

    override fun setLooping(isLooping: Boolean) {
        if (mInternalPlayer != null) mInternalPlayer!!.repeatMode =
            if (isLooping) Player.REPEAT_MODE_ALL else Player.REPEAT_MODE_OFF
    }

    override fun setSpeed(speed: Float) {
        val playbackParameters = PlaybackParameters(speed)
        mSpeedPlaybackParameters = playbackParameters
        if (mInternalPlayer != null) {
            mInternalPlayer!!.playbackParameters = playbackParameters
        }
    }

    override fun getSpeed(): Float {
        return if (mSpeedPlaybackParameters != null) {
            mSpeedPlaybackParameters!!.speed
        } else 1f
    }

    override fun getTcpSpeed(): Long {
        // no support
        return 0
    }

    override fun onPlaybackStateChanged(playbackState: Int) {
        if (eventListener == null) return
        if (mIsPreparing) {
            if (playbackState == Player.STATE_READY) {
                eventListener!!.onPrepared()
                eventListener!!.onInfo(IPlayer.MEDIA_INFO_RENDERING_START, 0)
                mIsPreparing = false
            }
            return
        }
        when (playbackState) {
            Player.STATE_BUFFERING -> eventListener!!.onInfo(
                IPlayer.MEDIA_INFO_BUFFERING_START,
                getBufferedPercentage()
            )
            Player.STATE_READY -> eventListener!!.onInfo(
                IPlayer.MEDIA_INFO_BUFFERING_END,
                getBufferedPercentage()
            )
            Player.STATE_ENDED -> eventListener!!.onCompletion()
            Player.STATE_IDLE -> {}
        }
    }

    override fun onPlayerError(error: PlaybackException) {
        if (eventListener != null) {
            eventListener!!.onError(
                PlayerException(
                    error
                )
            )
        }
    }

    override fun onVideoSizeChanged(videoSize: VideoSize) {
        if (eventListener != null) {
            eventListener!!.onVideoSizeChanged(videoSize.width, videoSize.height)
            if (videoSize.unappliedRotationDegrees > 0) {
                eventListener!!.onInfo(
                    IPlayer.MEDIA_INFO_VIDEO_ROTATION_CHANGED,
                    videoSize.unappliedRotationDegrees
                )
            }
        }
    }

    init {
        mAppContext = context.applicationContext
        mMediaSourceHelper = ExoMediaSourceHelper.getInstance(context)
    }
}